### DISCUSSION TITLE
WidgetRef of a not current Widget, is listen yet

### DISCUSSION THREAD

**Discussion by LaTrita97** [rrousselGit/riverpod#965] [Q&A] [answered]
I don't know if it's an issue, but by reading the documentation I don't find anything about this behavior.

Imagine that we have 2 pages, both are `ConsumerStatefulWidget`and inside their `build()` method they listen to a provider by
``` 
ref.listen(provider, (_, state) {
      //do something
    });
``` 
When the status changes, inside each `ref.listen` it is printed the name of the `Widget`.

**Widget 1** :
``` 
ref.listen(provider, (_, state) {
      print('widget_1');
    });
``` 
**Widget 2** :
``` 
ref.listen(provider, (_, state) {
      print('widget_2');
    });
``` 

When the app is started I'm in the `Widget1`, I pass on the `Widget2` by `Navigator.of(context).push(...)`, after that I change the status of `provider` from `Widget2` and in the console I see printed `widget_2` and also `widget_1`.

Why the **Widget1** is listening yet? Is normal?
How can I listen only to the changes of `Widget2` when I'm in the `Widget2`?
I would that **Widget1** not listen when I'm not in **Widget1**.

This example is a simple view of my problem, I need to listen to the same provider in both widgets.
Can someone help me to figure out why this happens and how to solve it?

Much appreciated.

**Comment by edsulaiman** [comment]
1. Why the Widget1 is listening yet? Is normal?
because `ref.listen` on Widget1 is still alive if the page is not popped. It's normal

2. How can I listen only to the changes of Widget2 when I'm in the Widget2?
I don't know about this but i think you can't do that. If your `ref.listen` do exactly the same thing, you can create a new provider for it
``` dart
final anotherProvider = Provider((ref) {
  ref.listen<int>(counterProvider, (int count) {
    print('The counter changed ${count}');
  });
  ...
});
```

(i rarely use `ref.listen`, but i have been using riverpod for a long time)

**Comment by rrousselGit** [comment]
Do you maybe want to use `Navigator.pushReplacement`? So that when showing Widget2, this pops Widget1

**Comment by LaTrita97** [comment, marked as answer]
From the responses I see that is not possible if I do not `pop` the widget that I want to exclude.
Thanks for the information and the help.

I use `ref.listen` for call methods depending on the state changes, I pass to the provider a class that has multiple states, and for each state, I have to call a different method.

Now like @rrousselGit suggests to me, I've used a `replace` rather `push`.

**Comment by NoName971** [comment]
I have the same problem here
I'm trying to centralize my auth logic in the same StateNotifier but I'm encountering the same issue
Is there an update about this ?
Or we should continue to use pop or replace or separate each function of the StateProvider in a dedicated provider ?
